code/model/modelread.cpp

```diff
@@ -4181,152 +4191,103 @@ void world_find_model_instance_point(vec3d *out, vec3d *world_pt, const polymode
 	}
 }
 
 /**
- * Same as find_submodel_instance_point_normal, except that this takes and
+ * Same as above, except that this takes and
  * returns matrices instead of normals.
  *  
- * Finds the current location and rotation (in the ship's frame of reference) of
+ * Finds the current location and rotation of
  * a submodel point, taking into account the rotations of the submodel and any
  * parent submodels it might have.
+ * 
+ * If the object parameters are supplied, the location will be in world coordinates;
+ * otherwise it will be in the model's reference frame.
  *
  * @param *outpnt Output point
  * @param *outorient Output matrix
  * @param model_instance_num Index into Polygon_model_instances
  * @param submodel_num The number of the submodel we're interested in
- * @param *submodel_pnt The point which's current position we want, in the submodel's frame of reference
- * @param *submodel_orient The local matrix which's current orientation in the ship's frame of reference we want
+ * @param *submodel_pnt The point whose current position we want, in the submodel's frame of reference
+ * @param *submodel_orient The local matrix whose current orientation in the ship's frame of reference we want
  */
-void find_submodel_instance_point_orient(vec3d *outpnt, matrix *outorient, const polymodel *pm, const polymodel_instance *pmi, int submodel_num, const vec3d *submodel_pnt, const matrix *submodel_orient)
+void model_instance_find_world_point_orient(vec3d *outpnt, matrix *outorient, const vec3d *submodel_pnt, const matrix *submodel_orient, const polymodel *pm, const polymodel_instance *pmi, int submodel_num, const matrix *objorient, const vec3d *objpos)
 {
+	vec3d pnt, tpnt;
+	matrix orient;
+	int mn;
 	Assert(pm->id == pmi->model_num);
-	*outorient = *submodel_orient;
-	vm_vec_zero(outpnt);
-
-	int mn = submodel_num;
-	while ( (mn >= 0) && (pm->submodel[mn].parent >= 0) ) {
-		vec3d offset = pm->submodel[mn].offset;
-
-		if ( mn == submodel_num) {
-			vec3d submodel_pnt_offset = *submodel_pnt;
 
-			vec3d tvec = submodel_pnt_offset;
-			vm_vec_unrotate(&submodel_pnt_offset, &tvec, &pmi->submodel[mn].canonical_orient);
-
-			matrix tnorm = *outorient;
-			vm_matrix_x_matrix(outorient, &tnorm, &pmi->submodel[mn].canonical_orient);
-
-			vm_vec_add2(&offset, &submodel_pnt_offset);
-		}
-
-		int parent_mn = pm->submodel[mn].parent;
-
-		vec3d tvec = offset;
-		vm_vec_unrotate(&offset, &tvec, &pmi->submodel[parent_mn].canonical_orient);
+	pnt = *submodel_pnt;
+	orient = *submodel_orient;
+	mn = submodel_num;
 
-		matrix tnorm = *outorient;
-		vm_matrix_x_matrix(outorient, &tnorm, &pmi->submodel[parent_mn].canonical_orient);
+	// instance up the tree for this point
+	while ( (mn >= 0) && (pm->submodel[mn].parent >= 0) ) {
+		vm_vec_unrotate(&tpnt, &pnt, &pmi->submodel[mn].canonical_orient);
+		vm_vec_add(&pnt, &tpnt, &pm->submodel[mn].offset);
 
-		vm_vec_add2(outpnt, &offset);
+		orient = orient * pmi->submodel[mn].canonical_orient;
```

Comment:
Just to comment on this:
The order for the matrix multiplications is changed here on purpose it seems, as the previous order would incorrectly apply the rotations to the stack. As we're going up the chain, and not down, this seems to change it to the correct order